Cabal-syntax/src/Distribution/Types/Dependency.hs

```diff
@@ -26,9 +33,55 @@ import Distribution.Types.LibraryName
 import Distribution.Types.PackageName
 import Distribution.Types.UnqualComponentName
 
+import qualified Distribution.Compat.CharParsing as P
 import qualified Distribution.Compat.NonEmptySet as NES
+import Distribution.ModuleName
 import qualified Text.PrettyPrint as PP
 
+data IsPrivate = Private PrivateAlias | Public deriving (Show, Ord, Read, Eq)
+
+data Dependencies = Dependencies {publicDependencies :: [Dependency], privateDependencies :: [PrivateDependency]} deriving (Eq, Show, Generic, Data)
+
+newtype PrivateAlias = PrivateAlias ModuleName deriving (Show, Eq, Generic, Data, Read, Ord)
+
+instance Pretty PrivateAlias where
+  pretty (PrivateAlias p) = pretty p
+
+instance Parsec PrivateAlias where
+  parsec = PrivateAlias <$> parsec
+
+-- | Construct a 'PrivateAlias' from a valid module name 'String'.
+--
+-- This is just a convenience function intended for valid module strings. It is
+-- an error if it is used with a string that is not a valid module name. If you
+-- are parsing user input then use 'Distribution.Text.simpleParse' instead.
+instance IsString PrivateAlias where
+  fromString = PrivateAlias . fromString
+
+data PrivateDependency = PrivateDependency {private_alias :: PrivateAlias, private_depends :: [Dependency]} deriving (Eq, Show, Generic, Data, Read, Ord)
+
+instance Parsec PrivateDependency where
+  parsec = do
+    alias <- parsec
+    P.spaces
+    _ <- P.string "with"
+    P.spaces
+    let parensLax p = P.between (P.char '(' >> P.spaces) (P.char ')' >> P.spaces) p
+    deps <- parensLax (parsecCommaList parsec)
+    return (PrivateDependency alias deps)
+
+instance Pretty PrivateDependency where
+  pretty (PrivateDependency alias deps) = PP.hsep [pretty alias, PP.text "with", PP.parens (PP.hsep (PP.punctuate PP.comma (map pretty deps)))]
+
+instance Semigroup Dependencies where
+  (Dependencies p1 pr1) <> (Dependencies p2 pr2) = Dependencies (p1 <> p2) (pr1 <> pr2)
```

Comment:
Not sure this does the right thing if the name of the private scopes are the same? Do we have a test for that?

Reply:
I see what you mean, we should probably group by PrivateAlias name and merge together the packages for private scopes with the same name within the same package.

I've pushed a test for this (`same-scope-name`).

Reply:
I suspect they should be merged yes, but I don't know what goes wrong if they aren't.

Reply:
The way things are, this is solved successfully:
```
cabal-version:   3.0
name:            pkgC
version:         0.1.0.0
build-type:      Simple

library
    private-build-depends: SameName with (libA == 0.1.0.0)
    build-depends:    base

executable myexe
    private-build-depends: SameName with (libA == 0.2.0.0)
    build-depends: base
```

I suspect grouping the private aliases here will make this no longer solve, which I feel is preferential within the same package (if you want different scopes, you could name them differently)

Reply:
Actually, this probably doesn't matter much (especially not for the above example).
In `qualifyDeps` we qualify a component's private dependencies with the component itself, meaning we'll differentiate between the component's private scopes's goals.

Perhaps `QualAlias` shouldn't be parametrised by a `Component` at all.

Reply:
> Perhaps QualAlias shouldn't be parametrised by a Component at all.
If we did this we'd no longer need a special `ConstraintScope` constructor for private scopes, we could reuse the qualifier one.